job-manager: add flux_job_wait()

garlick opened the pull request:
This PR adds a simple job synchronization mechanism.  One submits a job with a new flag, and then one can "wait' on the job and get back its status.   If the job is still active, "wait" blocks until it becomes inactive, as you might guess from the name.  The result of the job is distilled down to a boolean `success` and a human readable error string.

There is a new submit flag and a couple API calls:
```c
enum job_submit_flags {
    ...
    FLUX_JOB_WAITABLE = 4,      // flux_job_wait() will be used on this job
};

enum {
    FLUX_JOBID_ANY = 0xFFFFFFFFFFFFFFFF, // ~(uint64_t)0
};

/* Wait for jobid to enter INACTIVE state.
 * If jobid=FLUX_JOBID_ANY, wait for the next waitable job.
 * Fails with ECHILD if there is nothing to wait for.
 */
flux_future_t *flux_job_wait (flux_t *h, flux_jobid_t id);
int flux_job_wait_get_status (flux_future_t *f,
                              bool *success,
                              const char **errstr);
int flux_job_wait_get_id (flux_future_t *f, flux_jobid_t *id);
```

Only jobs submitted with the WAITABLE flag can be waited on.

Only the instance owner can use this, mainly to dodge the potential DoS that might arise if a user creates many "zombie" jobs, so potentially that could be changed if that concern is deemed low risk or we have an idea for mitigation.

The wait call can be made only once per job (it is destructive to the job completion state).

I had planned something like this all along, but after speaking with James Corbett the other day was reminded that it was still missing.

There is also a python binding, making e.g. this sort of thing possible:
```python
for i in range(njobs):
        jobid = job.submit(h, jobspec, flags=flux.constants.FLUX_JOB_WAITABLE)

for i in range(njobs):
    result = job.wait(h)  # missing jobid parameter == FLUX_JOBID_ANY
    if result.success:
        print("wait: {} Success".format(result.jobid))
    else:
        print("wait: {} Error: {}".format(result.jobid, result.errstr))
```

grondo commented:
What is the philosophical difference between `flux job wait ID` and `flux job wait-event ID clean`?

Is the new flux_job_wait() just meant for high throughput cases?

> and then one can "wait' on the job and get back the aggregate status code from the finish event (the largest of the status bytes returned by tasks).

Just to avoid confusion, the `finish` event has the largest status returned from the shells, not necessarily the tasks. The job-exec module doesn't have access to the wait status for tasks. (e.g. all tasks could have exited with status 0, but one or more shells may exit with a nonzero status if encountering a fatal error)

garlick commented:
> What is the philosophical difference between flux job wait ID and flux job wait-event ID clean

Interface wise, `flux_job_wait()` allows the user to synchronize on job completion and get the coarse-grained result in one call.  In theory `flux_job_wait()` should be more scable since it's co-located with the job manager where the state transitions occur, and thus has to do a lot less internally.

The motivation is to provide a usable API (ultimately in python) for a workflow front end to manage work as it moves through the system.  I like the similarity to `wait(2)` here, although (disclaimer) my background may bias me towards a familiar-to-me interface that others might find clunky.

> the finish event has the largest status returned from the shells, not necessarily the tasks

I'll clarify that in the commit message.  Pushing that and a few other changes:
* rebase on current master
* squash one commit I missed
* add coverage for exceptions
* add "zombies" to the `job-manager.list` output (per offline discussion)

garlick commented:
Just force pushed with (I think) some improvements:
* Returned payload consists of jobid, boolean success, and an error string.  (Before, it returned wait(2) status byte for user to decode)
* Don't fail the RPC if the job was terminated by exception (only fail if the wait request cannot be serviced)
* One can now wait on FLUX_JOBID_ANY and get the next waitable job that transitions to INACTIVE
* Minimal python binding (mimicking `job.submit()`)
* A couple simple python scripts demonstrating the API (run from sharness for good measure)

grondo commented:
> Don't fail the RPC if the job was terminated by exception (only fail if the wait request cannot be serviced)

Nice, I like this design *a lot* better!

grondo commented:
Looks like we have an immediate use case for this interface (#2558), so let me know how I can help get this one ready.

garlick commented:
I'll drop the tentative fixes pulled in from in and around #2553 and rebase on current master.

Note `job.wait()` has exactly theh same structure as `job.submit()` so will need to be included in the #2553 fix if this goes in first.

grondo commented:
> Note job.wait() has exactly theh same structure as job.submit() so will need to be included in the #2553 fix if this goes in first.

If the extra reference for Python futures is dropped as found by @SteVwonder, is that fix even still required? I guess it doesn't hurt to explicitly destroy futures when we know we are done with them...

garlick commented:
> If the extra reference for Python futures is dropped as found by @SteVwonder, is that fix even still required?

Oh possibly not.  There's still a test failure to run down with that one though.

I do think maybe one change is needed in this PR.  Currently the jobid is returned as the first item in a tuple by `job.wait()`.  However, `job.wait()` can be made with or without a jobid argument.  If without, FLUX_JOBID_ANY is assumed and then the returned jobid is useful.  If with, the returned jobid is redundant.  So maybe I should put it last so users can skip it if they are using the ANY pattern.

garlick commented:
> I do think maybe one change is needed in this PR.

Oh strike that.  Apparently leaving off one of the items in a returned tuple causes a `ValueError: too many values to unpack` exception, and it seems like conditionally returning two or three values depending on context might get a little confusing.  I guess this is ready for a review as is.

garlick commented:
I just dropped WIP from the description and updated the initial comment to reflect the end state of the proposal since it evolved a bit.

grondo commented:
> Apparently leaving off one of the items in a returned tuple causes a ValueError: too many values to unpack exception, and it seems like conditionally returning two or three values depending on context might get a little confusing.

I wonder if in this case it would be easier (for the user) to return a dict from the `job.wait()` function? I also notice that more recent Python has a `namedtuple` type (like a dict, but immutable), that might be more user friendly. Though, I'm not a python user so I'm not sure what would be most pythonic in this case.

e.g. instead of
```python
    jobid = job.submit(h, jobspec, flags=flux.constants.FLUX_JOB_WAITABLE)
    
    jobid, success, errstr = job.wait(h)
    if success:
        print("wait: {} Success".format(jobid))
    else:
        print("wait: {} Error: {}".format(jobid, errstr))
```

maybe
```python
    jobid = job.submit(h, jobspec, flags=flux.constants.FLUX_JOB_WAITABLE)

    result = job.wait(h)
    if result.success then:
        print("wait: {} Success".format(result.id))
    else
        print("wait: {} Error: {}".format(result.id, result.errstr))

```

@andre-merzky, any comment since you might be the first user of this interface?

andre-merzky commented:
@grondo : thanks for the ping!  The named tuple looks great to me, really: its readable, easy to use, and gives a consistent API in both cases.

grondo commented:
@garlick, this quick test seemed to work:
```diff
diff --git a/src/bindings/python/flux/job.py b/src/bindings/python/flux/job.py
index 0fc3973..ecf152e 100644
--- a/src/bindings/python/flux/job.py
+++ b/src/bindings/python/flux/job.py
@@ -11,6 +11,7 @@
 import errno
 
 import six
+import collections
 
 from flux.wrapper import Wrapper
 from flux.util import check_future_error
@@ -58,6 +59,7 @@ def wait_async(flux_handle, jobid=lib.FLUX_JOBID_ANY):
     future_handle = RAW.wait(flux_handle, jobid)
     return Future(future_handle)
 
+JobWaitResult = collections.namedtuple('JobWaitResult', 'jobid, success, errstr')
 
 @check_future_error
 def wait_get_status(future):
@@ -69,8 +71,7 @@ def wait_get_status(future):
     jobid = ffi.new("flux_jobid_t[1]")
     RAW.wait_get_id(future, jobid)
     RAW.wait_get_status(future, success, errstr)
-    return int(jobid[0]), bool(success[0]), ffi.string(errstr[0]).decode("utf-8
-
+    return JobWaitResult(int(jobid[0]), bool(success[0]), ffi.string(errstr[0]))
 
 def wait(flux_handle, jobid=lib.FLUX_JOBID_ANY):
     future = wait_async(flux_handle, jobid)

```

as a nice side effect, the namedtuple can still be unpacked as a normal tuple, so existing usage wouldn't break. However, callers can do something like this:

```python
while count > 0:
    r = job.wait(h)
    if not r.success:
        print("wait: {} Error: {}".format(r.jobid, r.errstr))
    count -= 1
```

garlick commented:
I really like that change!  Thanks guys.  Will force a push shortly.

garlick commented on src/cmd/flux-mini.py:
```diff
@@ -315,6 +315,9 @@ def create_parser(self):
         parser.add_argument("--debug", action="store_true", help="Set job debug flag")
+        parser.add_argument(
+            "--waitable", action="store_true", help="Set job waitable flag"
+        )
```
Sure, let's fix that now.  Better if options are similar across the different commands.

grondo commented:
Feel free to set `merge-when-passing` when you are ready. It will be nice to get this in now even if it may require minor tweaks going forward, so that we can get some real user feedback.

garlick commented:
I went ahead and added the flags option to flux mini run|submit:
```
     --flags=FLAGS       Set comma separated list of job submission flags.
                         Possible flags: debug, waitable
```
and also rebased on master.  Please let me know if that option looks OK.

grondo commented:
Looks good! I Added merge-when-passing label.

garlick commented:
Thanks!
